src/mame/drivers/firebeat.cpp

```diff
@@ -865,368 +1127,381 @@ WRITE_LINE_MEMBER(firebeat_state::spu_ata_dmarq)
-	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_BUTTON2 )            // Switch 2
-	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_BUTTON3 )            // Switch 3
-	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_BUTTON4 )            // Switch 4
-	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_BUTTON5 )            // Switch 5
-	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON6 )            // Switch 6
-	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_BUTTON7 )            // Switch 7
-	PORT_BIT( 0x80, IP_ACTIVE_LOW, IPT_BUTTON8 )            // Switch 8
+/*****************************************************************************
+* ParaParaParadise / ParaParaDancing
+******************************************************************************/
+void firebeat_ppp_state::firebeat_ppp(machine_config &config)
+{
+	firebeat(config);
+	m_maincpu->set_addrmap(AS_PROGRAM, &firebeat_ppp_state::firebeat_ppp_map);
+}
 
-	PORT_START("IN1")
-	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_BUTTON9 )            // Switch 9
-	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_UNKNOWN )
-	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_COIN1 )              // Coin
-	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_UNKNOWN )
-	PORT_SERVICE_NO_TOGGLE( 0x10, IP_ACTIVE_LOW)            // Test
-	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_UNKNOWN )
-	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_SERVICE ) PORT_NAME("Service") PORT_CODE(KEYCODE_7)      // Service
-	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_UNKNOWN )
+void firebeat_ppp_state::init_ppp()
+{
+	init_firebeat();
+	init_lights(write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output_ppp_w)), write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output2_ppp_w)), write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output3_ppp_w)));
 
-INPUT_PORTS_END
+	m_cabinet_info = 8;
+}
 
-INTERRUPT_GEN_MEMBER(firebeat_state::firebeat_interrupt)
+void firebeat_ppp_state::init_ppd()
 {
-	// IRQs
-	// IRQ 0: VBlank
-	// IRQ 1: Extend board IRQ
-	// IRQ 2: Main board UART
-	// IRQ 3: SPU mailbox interrupt
-	// IRQ 4: ATA
+	init_firebeat();
+	init_lights(write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output_ppp_w)), write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output2_ppp_w)), write32s_delegate(*this, FUNC(firebeat_ppp_state::lamp_output3_ppp_w)));
 
-	device.execute().set_input_line(INPUT_LINE_IRQ0, ASSERT_LINE);
+	m_cabinet_info = 9;
 }
 
-WRITE_LINE_MEMBER(firebeat_state::gcu0_interrupt)
+void firebeat_ppp_state::firebeat_ppp_map(address_map &map)
 {
-	m_maincpu->set_input_line(INPUT_LINE_IRQ0, state);
+	firebeat_map(map);
+	map(0x7d000340, 0x7d00035f).r(FUNC(firebeat_ppp_state::sensor_r));
 }
 
-WRITE_LINE_MEMBER(firebeat_state::gcu1_interrupt)
+uint16_t firebeat_ppp_state::sensor_r(offs_t offset)
 {
-	m_maincpu->set_input_line(INPUT_LINE_IRQ0, state);
+	switch (offset) {
+		case 0: return ioport("SENSOR1")->read() | 0x0100;
+		case 1: return ioport("SENSOR2")->read() | 0x0100;
+		case 2: return ioport("SENSOR3")->read() | 0x0100;
+		case 3: return ioport("SENSOR4")->read() | 0x0100;
```

Comment:
Please use `required_ioport_array` (or possibly `optional_ioport_array`) for these.